src/Parameter.php

```diff
@@ -652,4 +707,127 @@ public function has($var)
         }
         return isset($this->{$var}) && !empty($this->{$var});
     }
+
+    /**
+     * Filters the given data using filter methods specified in the config.
+     *
+     * If $stage is provided, only filters that apply to the provided filter
+     * stage will be invoked. To preserve legacy behavior, filters that do not
+     * specify a stage are implicitly invoked only in the pre-validation stage.
+     *
+     * @param mixed $value The value to filter.
+     * @param string $stage An optional specifier of what filter stage to
+     *     invoke. If null, then all filters are invoked no matter what stage
+     *     they apply to. Otherwise, only filters for the specified stage are
+     *     invoked.
+     *
+     * @return mixed The filtered value.
+     */
+    private function invokeCustomFilters($value, $stage) {
+        $filteredValue = $value;
+
+        foreach ($this->filters as $filter) {
+            if (is_array($filter)) {
+                $filteredValue =
+                    $this->invokeComplexFilter($filter, $value, $stage);
+            } else {
+                $filteredValue =
+                    $this->invokeSimpleFilter($filter, $value, $stage);
+            }
+        }
+
+        return $filteredValue;
+    }
+
+    /**
+     * Invokes a filter that uses value substitution and/or should only be
+     * invoked for a particular filter stage.
+     *
+     * If $stage is provided, and the filter specifies a stage, it is not
+     * invoked unless $stage matches the stage the filter indicates it applies
+     * to. If the filter is not invoked, $value is returned exactly as it was
+     * provided to this method.
+     *
+     * To preserve legacy behavior, if the filter does not specify a stage, it
+     * is implicitly invoked only in the pre-validation stage.
+     *
+     * @param array $filter Information about the filter to invoke.
+     * @param mixed $value The value to filter.
+     * @param string $stage An optional specifier of what filter stage to
+     *     invoke. If null, then the filter is invoked no matter what stage it
+     *     indicates it applies to. Otherwise, the filter is only invoked if it
+     *     matches the specified stage.
+     *
+     * @return mixed The filtered value.
+     */
+    private function invokeComplexFilter(array $filter, $value, $stage) {
+        if (isset($filter['stage'])) {
+            $filterStage = $filter['stage'];
+        } else {
+            $filterStage = self::FILTER_STAGE_AFTER_VALIDATION;
```

Comment:
This default is fine for filters on request payloads/parameters, but wrong for filters on response values. For responses, the default filter stage should be `FILTER_STAGE_RESPONSE_WIRE` if not specified.

When a simple filter is used, the logic is correct because the filter is invoked for all stages of processing except `FILTER_STAGE_AFTER_VALIDATION`. But if that simple filter is rewritten as a complex filter (by specifying the filter as an array with a `method` key), the filter will not work properly without `stage` key.